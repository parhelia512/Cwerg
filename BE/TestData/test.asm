# comment
;;;  comment

.num four pos 4
.num pointone flt .1111
    
    
.mem global1 four ro
	.data 16 [0 0 1 1 2 2 3 3]
	.data 1 "hello world\0"
    
.mem global2 four rw
    
.fun fun1 normal [s32 rout1 u8 rout2] = [s8 rin1 u8 rin2]
    .stk slot1 4 8
    .num minus_one neg -1	
    .num plus_one pos 1	
    .reg [r01 r02 r03] s8
    .reg [r11 r12 r13] s32
    .reg [r21 r22 r23] a32
    .reg [r31 r32 r33] f32
    .reg [r41] a32
    .reg [r91] c32

.bbl start
    xor rout2 rin2 rin2
    add r01 = r02 r03
    add r11 = r12 r13
    add r31 = r32 r33
    add.i r01 = r01 minus_one
    mov.stk r41 = slot1
    mov.mem r41 = global1
    mov r02 = r03
    bra start
    
.bbl loop
    beq start r22 r32
    beq next r22 r32

.bbl next
    mov.fun r91 = fun1
    add r21 = r21 r11
    sub r11 r21 r22
    bsr fun2 [rout1 rout2] = [rin1 rin2]
    ret 
	
.fun fun2 normal [s32 rout1 u8 rout2] = [s8 rin1 u8 rin2]
    .reg [r01 r02 r03] s8
