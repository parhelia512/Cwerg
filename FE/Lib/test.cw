-- test helpers macros
--
-- This intentionally does not import the fmt module to keep
-- the footprint/dependencies small.
-- (We may change our mind on this.)
--
module:

import os

macro SysPrint# STMT_LIST($msg EXPR)[$msg_eval]:
    mlet $msg_eval span(u8) = $msg
    do os::write(unwrap(os::Stdout), front($msg_eval), len($msg_eval))

pub macro Success# STMT()[]:
    SysPrint#("OK\n")

-- The two scalar arguments must be the same
--
-- Both must have derivable types as we use `auto`
pub macro AssertEq# STMT_LIST($e_expr EXPR, $a_expr EXPR)[$e_val, $a_val]:
    mlet $e_val = $e_expr
    mlet $a_val = $a_expr
    if $e_val != $a_val:
        SysPrint#("AssertEq failed: ")
        SysPrint#(stringify($e_expr))
        SysPrint#(" VS ")
        SysPrint#(stringify($a_expr))
        SysPrint#("\n")
        trap

-- The two scalar arguments must be the same
--
-- Both must have derivable types as we use `auto`
pub macro AssertNe# STMT_LIST($e_expr EXPR, $a_expr EXPR)[$e_val, $a_val]:
    mlet $e_val = $e_expr
    mlet $a_val = $a_expr
    if $e_val == $a_val:
        SysPrint#("AssertNe failed: ")
        SysPrint#(stringify($e_expr))
        SysPrint#(" VS ")
        SysPrint#(stringify($a_expr))
        SysPrint#("\n")
        trap

-- First argument must have the type denoted by the second
pub macro AssertIs# STMT_LIST($expr EXPR, $type TYPE)[]:
    if is($expr, $type):
    else:
        SysPrint#("AssertIs failed: ")
        SysPrint#(stringify(type_of($expr)))
        SysPrint#(" VS ")
        SysPrint#(stringify($type))
        SysPrint#("\n")
        trap

-- The two arguments must type derivable
pub macro AssertSliceEq# STMT_LIST($e_expr EXPR, $a_expr EXPR)[
        $e_val, $a_val, $i]:
    mlet $e_val = $e_expr
    mlet $a_val = $a_expr
    AssertEq#(len($e_val), len($a_val))
    for $i = 0, len($a_val), 1:
        AssertEq#(pinc(front($e_val), $i)^, pinc(front($a_val), $i)^)

-- The first two arguments must derivable types as we use `auto`
pub macro AssertApproxEq# STMT_LIST($e_expr EXPR, $a_expr EXPR, $epsilon EXPR)[
        $e_val, $a_val]:
    mlet $e_val = $e_expr
    mlet $a_val = $a_expr
    if abs($e_val - $a_val) >= $epsilon:
        SysPrint#("AssertApproxEq failed: ")
        SysPrint#(stringify($e_expr))
        SysPrint#(" VS ")
        SysPrint#(stringify($a_expr))
        SysPrint#("\n")
        trap

pub macro AssertEqR64# STMT_LIST($e_expr EXPR, $a_expr EXPR)[$e_val, $a_val]:
    mlet $e_val r64 = $e_expr
    mlet $a_val r64 = $a_expr
    if bitwise_as($e_val, u64) != bitwise_as($a_val, u64):
        SysPrint#("AssertEq failed: ")
        SysPrint#(stringify($e_expr))
        SysPrint#(" VS ")
        SysPrint#(stringify($a_expr))
        SysPrint#("\n")
        trap

pub macro AssertNeR64# STMT_LIST($e_expr EXPR, $a_expr EXPR)[$e_val, $a_val]:
    mlet $e_val r64 = $e_expr
    mlet $a_val r64 = $a_expr
    if bitwise_as($e_val, u64) == bitwise_as($a_val, u64):
        SysPrint#("AssertEq failed: ")
        SysPrint#(stringify($e_expr))
        SysPrint#(" VS ")
        SysPrint#(stringify($a_expr))
        SysPrint#("\n")
        trap

-- The first two arguments must type derivable
pub macro AssertSliceApproxEq# STMT_LIST($e_expr EXPR, $a_expr EXPR, $epsilon EXPR)[
        $e_val, $a_val, $i]:
    mlet $e_val = $e_expr
    mlet $a_val = $a_expr
    AssertEq#(len($e_val), len($a_val))
    for $i = 0, len($a_val), 1:
        AssertApproxEq#(
                pinc(front($e_val), $i)^, pinc(front($a_val), $i)^, $epsilon)

--
pub macro AssertTrue# STMT_LIST($e_expr EXPR)[]:
    if $e_expr:
    else:
        SysPrint#("AssertTrue failed: ")
        SysPrint#(stringify($e_expr))
        SysPrint#("\n")
        trap

--
pub macro AssertFalse# STMT_LIST($e_expr EXPR)[]:
    if $e_expr:
        SysPrint#("AssertFalse failed: ")
        SysPrint#(stringify($e_expr))
        SysPrint#("\n")
        trap

--
pub macro AssertUnreachable# STMT_LIST()[]:
    SysPrint#("AssertUnreachable\n")
    trap
