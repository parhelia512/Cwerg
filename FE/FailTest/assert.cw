module:

import test

fun main(argc s32, argv ^^u8) s32:
    test::AssertEq#(0_uint, 1_uint)
    test::Success#()
    return 0