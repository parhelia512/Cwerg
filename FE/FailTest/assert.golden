AssertEq failed in assert.cw(7): [ValNum] 0_uint VS [ValNum] 1_uint
